### PR TITLE
:sparkles: Add tla2528 driver

### PR DESCRIPTION
**Pull request by MichaelYKersey**
The tla2528 is an i2c-based GPIO expander and ADC mux. It has 8 pins that can each be independently configured and operated as GPIO or ADC. This is a port of the tla2528 driver and adapters from the [SJSU robotics team](https://github.com/SJSURoboticsTeam/urc-control-systems-2024/pull/12). 

Features
- ADC readings
- Digital input readings (bus & individual)
- Digital output writing (bus & individual) (push-pull & open-drain)
- Adapters for each

Missing/future/unimplemented features:
- TLA reset on object creation
- ADC 16-bit precision (currently 12-bit, but chip allows for more)
- ADC channel verification (redundancy)

### CODE REVIEW

**Comment by MichaelYKersey** [comment]
currently, the reset command is declared as a private method but is unimplemented. Should it be removed? or is it better to leave it there for the binary interface consistency (I'm not sure how ABIs work)

File: demos/applications/tla2528_analog_input.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#include <libhal-util/bit.hpp>
+#include <libhal-util/i2c.hpp>
+#include <libhal-util/serial.hpp>
+#include <libhal-util/steady_clock.hpp>
+#include <libhal/input_pin.hpp>
+#include <libhal/units.hpp>
+#include <libhal-expander/tla2528.hpp>
+
+#include <libhal-expander/tla2528_adapters.hpp>
+#include <resource_list.hpp>
```
Comment (kammce) [inline demos/applications/tla2528_analog_input.cpp]:
```suggestion
#include <array>

#include <libhal-util/bit.hpp>
#include <libhal-util/i2c.hpp>
#include <libhal-util/serial.hpp>
#include <libhal-util/steady_clock.hpp>
#include <libhal/input_pin.hpp>
#include <libhal/units.hpp>
#include <libhal-expander/tla2528.hpp>
#include <libhal-expander/tla2528_adapters.hpp>

#include <resource_list.hpp>
```

Suggestion made based on https://libhal.github.io/4.1/contributor_guide/style/#s143-include-order.

File: demos/applications/tla2528_analog_input.cpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+                                   make_adc(adc_mux, 6), make_adc(adc_mux, 7) };
+
+  while (true) {
+    hal::print(terminal, "\nvalues:\n");
+    for (int i = 0; i < 8; i++) {
+      hal::print<64>(terminal, "%d:%f\n", i, adcs[i].read());
+    }
+    hal::delay(steady_clock, 500ms);
+  }
+}
```
Comment (kammce) [inline demos/applications/tla2528_analog_input.cpp]: All files should end with a newline. I'm adding a patch to have CI check demos directory files as well.

File: demos/applications/tla2528_input_pin.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#include <libhal-util/bit.hpp>
+#include <libhal-util/i2c.hpp>
+#include <libhal-util/serial.hpp>
+#include <libhal-util/steady_clock.hpp>
+#include <libhal/input_pin.hpp>
+#include <libhal/units.hpp>
+#include <libhal-expander/tla2528.hpp>
+#include <libhal-expander/tla2528_adapters.hpp>
+#include <resource_list.hpp>
+
```
Comment (kammce) [inline demos/applications/tla2528_input_pin.cpp]:
```suggestion
#include <array>

#include <libhal-util/bit.hpp>
#include <libhal-util/i2c.hpp>
#include <libhal-util/serial.hpp>
#include <libhal-util/steady_clock.hpp>
#include <libhal/input_pin.hpp>
#include <libhal/units.hpp>
#include <libhal-expander/tla2528.hpp>
#include <libhal-expander/tla2528_adapters.hpp>

#include <resource_list.hpp>

```

File: include/libhal-expander/tla2528_adapters.hpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * @param p_settings input pin settings, default is no resistor
+ *
+ * @throws hal::argument_out_of_domain - if p_channel out of range (>7)
+ *
+ * @throws hal::resource_unavailable_try_again - if an adapter has already been
+ * been made for the pin
+ *
+ * @throws hal::no_such_device - no device responded on i2c bus
+ */
```
Comment (kammce) [inline include/libhal-expander/tla2528_adapters.hpp]:
```suggestion
/**
 * @brief create a hal::input_pin driver using the tla2528 driver
 *
 * @param p_channel pin acting as an input pin
 * @param p_settings input pin settings, default is no resistor
 * @throws hal::argument_out_of_domain - if p_channel out of range (>7)
 * @throws hal::resource_unavailable_try_again - if an adapter has already been
 * been made for the pin
 * @throws hal::no_such_device - no device responded on i2c bus
 */
```

I find the spacing harder to read than keeping everything together. Can you make this change to other comments like this as well.

File: include/libhal-expander/tla2528_adapters.hpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+tla2528_input_pin make_input_pin(tla2528& p_tla2528,
+                                 hal::byte p_channel,
+                                 hal::input_pin::settings const& p_settings);
```
Comment (kammce) [inline include/libhal-expander/tla2528_adapters.hpp]:
```suggestion
                                 hal::input_pin::settings const& p_settings = {.resistor = none});
```

May be a nice QOL feature to have this have a default settings parameter vs having the user have to supply each.

File: src/tla2528.cpp [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  auto_seq_ch_sel = 0x12
+};
+}  // namespace
+namespace hal::expander {
+
+tla2528::tla2528(hal::i2c& p_i2c, hal::byte p_i2c_address)
+  : m_i2c_bus(p_i2c)
+  , m_i2c_address(p_i2c_address)
+{
+  // TODO: reset command
```
Comment (kammce) [inline src/tla2528.cpp]:
```suggestion
  // TODO(#issue-number): Add reset command
```

Each commented TODO should also have an issue and issue number accompanied with it so its easier to follow the technical debt. Please create an issue and label it with technical debt. If you cannot label it due to permission reasons, say so in a reply and I'll label the issue for you.

File: include/libhal-expander/tla2528.hpp [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+// adapters
+class tla2528_adc;
+class tla2528_input_pin;
+class tla2528_output_pin;
+
+/**
+ * @brief tla2528 is a gpio expander & adc mux driver
+ *
+ * tla2528 has 8 pins which can be independently operated as an adc,
+ * digital input, and digital out over i2c. The i2c address is configured by
```
Comment (kammce) [inline include/libhal-expander/tla2528.hpp]:
```suggestion
 * input pin, and digital pin over i2c. The i2c address is configured by
```

Might as well use the same terminology we use for the interfaces.

File: include/libhal-expander/tla2528.hpp [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+   * @param p_high the output level of the pin, true is high, false is low.
+   *
+   * @throws hal::argument_out_of_domain - if p_channel out of range (>7)
+   *
+   * @throws hal::no_such_device - no device responded on i2c bus
+   *
+   * The device will write to a register that caches the desired output state.
+   * When a pin is in a digital output mode it will reference the desired state
+   * cache. If the pin is not set to digital output it will just be stored.
+   */
```
Comment (kammce) [inline include/libhal-expander/tla2528.hpp]:
```suggestion
  /**
   * @brief set digital output level pin
   *
   * The device will write to a register that caches the desired output state.
   * When a pin is in a digital output mode it will reference the desired state
   * cache. If the pin is not set to digital output it will just be stored.
   *
   * @param p_channel pin to set output
   * @param p_high the output level of the pin, true is high, false is low.
   * @throws hal::argument_out_of_domain - if p_channel out of range (>7)
   * @throws hal::no_such_device - no device responded on i2c bus
   */
```

Put the description after the brief then define the parameters and throwables. But I also don't think the description makes sense to a reader. I think the underlying implementation, in this case, isn't important enough to spell out to the API user. In other cases I could see a rationale for stating the implementation details in order to warn a user about a potentially expensive operation. But in this case, I don't see it as very useful.

File: include/libhal-expander/tla2528.hpp [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+   * @param p_mode tla2528::pin_mode enum of desired pin mode
+   *
+   * @param p_channel which pin to configure
+   *
+   * @throws hal::argument_out_of_domain - if p_channel out of range (>7)
+   *
+   * @throws hal::resource_unavailable_try_again - if adapters are made for a
+   * pin an exception may be thrown to prevent invalid behavior
+   *
+   * @throws hal::no_such_device - no device responded on i2c bus
```
Comment (kammce) [inline include/libhal-expander/tla2528.hpp]:
```suggestion
```

I think its fine to only list the exceptions that this driver explicity emits vs what the underlying drivers use. Yes, an i2c driver emits a "no such device" exception at every call that interacts with i2c, but tracking all of these exception through comments is tedious and error prone. We are working on tools to make this easier. The developer using this driver should also implicitly know that i2c exceptions can be emitted from a driver that use i2c. So we can remove all of these throws statements for i2c specific things.

File: include/libhal-expander/tla2528.hpp [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+   *
+   * If the pin is not set to output pin the returned state will be used once it
+   * changes to an output pin.
+   */
+  hal::byte get_digital_bus_out();
+
+  /**
+   * @brief read the digital level of a pins
+   *
+   * @return if the pin's digital read value is high.
```
Comment (kammce) [inline include/libhal-expander/tla2528.hpp]:
```suggestion
   * @return true if the pin's digital read value is high.
```

File: include/libhal-expander/tla2528.hpp [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+   *
+   * @throws hal::argument_out_of_domain - if p_channel out of range. (>7)
+   *
+   * @throws hal::no_such_device - no device responded on i2c bus
+   *
+   * If a pin is not set to digital input or output the returned value may not
+   * correlate with the true value.
```
Comment (kammce) [inline include/libhal-expander/tla2528.hpp]: It would be nice to put this in the returns block vs in the description so its clear when the reader reads the return documentation.

Comment (kammce) [reply include/libhal-expander/tla2528.hpp]: Same goes for the others.

File: src/tla2528.cpp [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+  hal::write(m_i2c_bus, m_i2c_address, cmd_buffer);
+}
+
+void tla2528::set_pin_mode(pin_mode p_mode, hal::byte p_channel)
+{
+  throw_if_invalid_channel(p_channel);
+  std::array<hal::byte, 5> data_buffer;
+  std::array<hal::byte, 2> read_cmd_buffer = {
+    op_codes::continuous_register_read, register_addresses::pin_cfg
+  };
```
Comment (kammce) [inline src/tla2528.cpp]:
```suggestion
  constexpr std::array<hal::byte, 2> read_cmd_buffer = {
    op_codes::continuous_register_read, register_addresses::pin_cfg
  };
```

Should be able to make arrays like these that do not change into constexpr objects.

File: src/tla2528.cpp [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+  hal::byte pin_cfg_reg = data_buffer[0];
+  hal::byte gpio_cfg_reg = data_buffer[2];
+  hal::byte gpo_drive_cfg_reg = data_buffer[4];
+
+  hal::bit_mask channel_mask = hal::bit_mask::from(p_channel);
+  if (p_mode == pin_mode::digital_output_push_pull ||
+      p_mode == pin_mode::digital_output_open_drain) {
+    if (hal::bit_extract(channel_mask, m_object_created) &&
+        !(hal::bit_extract(channel_mask, pin_cfg_reg) ||
+          hal::bit_extract(channel_mask, gpio_cfg_reg))) {
```
Comment (kammce) [inline src/tla2528.cpp]:
```suggestion
    if (hal::bit_extract(channel_mask, m_object_created) &&
        !(hal::bit_extract(channel_mask, pin_cfg_reg) ||
          hal::bit_extract(channel_mask, gpio_cfg_reg))) {
```

This logic is hard to follow. The `m_object_created` makes sense, that this is checking if the adaptor has been created but the other flags are not obvious. What you can do is break these up into bools that represent what you mean to express here. Something like

```C++
bool object_already_created = hal::bit_extract(channel_mask, m_object_created);
// etc for what the other ones mean.
bool misconfigured = not hal::bit_extract(channel_mask, pin_cfg_reg) || hal::bit_extract(channel_mask, gpio_cfg_reg);
if (object_already_created && misconfigured) { /* ... */ }
```

Something like that.

File: src/tla2528.cpp [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+  };
+  hal::write(m_i2c_bus, m_i2c_address, write_cmd_buffer);
+}
+
+void tla2528::set_digital_bus_out(hal::byte p_values)
+{
+  m_gpo_value = p_values;
+  std::array<hal::byte, 3> cmd_buffer = { op_codes::single_register_write,
+                                          register_addresses::gpo_value,
+                                          m_gpo_value };
```
Comment (kammce) [inline src/tla2528.cpp:125]:
```suggestion
  constexpr std::array<hal::byte, 3> cmd_buffer = { op_codes::single_register_write,
                                          register_addresses::gpo_value,
                                          m_gpo_value };
```

File: src/tla2528.cpp [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+bool tla2528::get_digital_in(hal::byte p_channel)
+{
+  throw_if_invalid_channel(p_channel);
+  return hal::bit_extract(hal::bit_mask::from(p_channel), get_digital_bus_in());
+}
+
+float tla2528::get_analog_in(hal::byte p_channel)
+{
+  set_analog_channel(p_channel);
+  // TODO: look into averaging & channel validation
```
Comment (kammce) [inline src/tla2528.cpp]:
```suggestion
  // TODO(#issue): look into averaging & channel validation
```

Make an issue for this.

File: src/tla2528_adapters.cpp [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+  tla2528& p_tla2528,
+  hal::byte p_channel,
+  hal::output_pin::settings const& p_settings)
+  : m_tla2528(&p_tla2528)
+  , m_channel(p_channel)
+{
+  if (hal::bit_extract(hal::bit_mask::from(m_channel),
+                       m_tla2528->m_object_created)) {
+    throw hal::resource_unavailable_try_again(this);
+  }
```
Comment (kammce) [inline src/tla2528_adapters.cpp]: You use this same logic a few times in the adaptors. Would make sense to make a helper function for this.

File: src/tla2528.cpp (hunk 6 above)
Comment (kammce) [inline src/tla2528.cpp]:
```suggestion
  // TODO(#issue-number): reset command
```

Make an technical debt issue for this. Also you might as well call reset() here and define it with the TODO. That way the code is already in place and the empty function has the todo embedded in it.

File: demos/applications/tla2528_output_pin.cpp [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    make_output_pin(gpo_expander, 2, output_pin_config),
+    make_output_pin(gpo_expander, 3, output_pin_config),
+    make_output_pin(gpo_expander, 4, output_pin_config),
+    make_output_pin(gpo_expander, 5, output_pin_config),
+    make_output_pin(gpo_expander, 6, output_pin_config),
+    make_output_pin(gpo_expander, 7, output_pin_config)
+  };
+
+  hal::byte counter =
+    0;  // output counts in binary to go though all out put combinations
```
Comment (kammce) [inline demos/applications/tla2528_output_pin.cpp]:
```suggestion
  // output counts in binary to go though all out put combinations
  hal::byte counter = 0;  
```

The newline 0 looks really odd. You can put the comment one line above.

File: demos/applications/tla2528_input_pin.cpp (hunk 3 above)
Comment (kammce) [reply demos/applications/tla2528_input_pin.cpp]: Do this everywhere else as well.

File: src/tla2528_adapters.cpp [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+  driver_configure(p_settings);
+  hal::bit_modify(m_tla2528->m_object_created)
+    .set(hal::bit_mask::from(m_channel));
+}
+tla2528_output_pin::~tla2528_output_pin()
+{
+  hal::bit_modify(m_tla2528->m_object_created)
+    .set(hal::bit_mask::from(m_channel));
```
Comment (kammce) [inline src/tla2528_adapters.cpp]:
```suggestion
    .clear(hal::bit_mask::from(m_channel));
```
Shouldn't this be clear?

Comment (kammce) [reply src/tla2528_adapters.cpp]: Do the same for the others.

**Review by kammce — CHANGES_REQUESTED** [review]
Great work! Much of my comments have to do with formatting and documentation. The rest mostly looks good.

**Comment by kammce** [comment]
> currently, the reset command is declared as a private method but is unimplemented. Should it be removed? or is it better to leave it there for the binary interface consistency (I'm not sure how ABIs work)

See my comment on what you should do about reset. As for ABI, class functions do not effect the ABI of a function so nothing to worry about from that angle.

File: src/tla2528.cpp (hunk 6 above)
Comment (MichaelYKersey) [reply src/tla2528.cpp]: I made the issue(s), I don't have permission to label them.

File: include/libhal-expander/tla2528.hpp (hunk 7 above)
Comment (MichaelYKersey) [reply include/libhal-expander/tla2528.hpp]: Which interface are you referring to?

File: src/tla2528.cpp (hunk 13 above)
Comment (MichaelYKersey) [reply src/tla2528.cpp]: Tried to do what you suggested, look it over.

File: include/libhal-expander/tla2528.hpp (hunk 8 above)
Comment (MichaelYKersey) [reply include/libhal-expander/tla2528.hpp]: Should I remove the descriptions, improve the clarity, or was it just unnecessary work that can be left alone?

Either way, I changed the order as requested.

File: include/libhal-expander/tla2528.hpp (hunk 7 above)
Comment (kammce) [reply include/libhal-expander/tla2528.hpp]: `hal::input_pin` and `hal::output_pin` so you can use the terms input pin and output pin which assume they are digital.

File: include/libhal-expander/tla2528.hpp (hunk 8 above)
Comment (kammce) [reply include/libhal-expander/tla2528.hpp]: Reading it over again, you can probably eliminate that section. If anything may be useful to be as a comment in the source file for readers looking to understand the code. I'll leave it to you if you want to place this somewhere in the source.

**Comment by kammce** [comment]
I can perform a other review once CI has passed. I recently re-enabled clang-tidy static analysis which is why there are so many new errors popping up.

File: src/tla2528.cpp (hunk 13 above)
Comment (MichaelYKersey) [reply src/tla2528.cpp]: approved in-person

File: include/libhal-expander/tla2528.hpp (hunk 7 above)
Comment (MichaelYKersey) [reply include/libhal-expander/tla2528.hpp]: Updated, look over it over

File: src/tla2528.cpp [hunk 19]
```diff
@@ -108,72 +114,75 @@ void tla2528::set_pin_mode(pin_mode p_mode, hal::byte p_channel)
...
+  // TODO(#8): look into averaging & channel validation
   std::array<hal::byte, 2> data_buffer;
   std::array<hal::byte, 1> cmd_buffer = { op_codes::single_register_read };
   hal::write_then_read(m_i2c_bus, m_i2c_address, cmd_buffer, data_buffer);
 
   // Take 12 bit number stored in first 12 bits of 2 bytes and converting to 16
   // bit num by shifting 4 bit right (See Figure 25 on datasheet)
   uint16_t data = data_buffer[0] << 4 | data_buffer[1] >> 4;
-  return data / 4095.0;
+  return (float)data / (float)4095.0;
```
Comment (kammce) [inline src/tla2528.cpp]: Use C++ style casts (clang-tidy should have caught this). 
```suggestion
  return static_cast<float>(data) / 4095.0f;
```

**Review by kammce — APPROVED** [review]
Looks good! One last little fix in the form of a C style cast and we can submit this.

File: include/libhal-expander/tla2528.hpp (hunk 7 above)
Comment (MichaelYKersey) [reply include/libhal-expander/tla2528.hpp]: marking as resolved based on general comment
"Looks good! One last little fix in the form of a C style cast and we can submit this."
